overlord/devicestate/devicestate_gadget_test.go

```diff
@@ -1386,11 +1425,161 @@ func (s *deviceMgrGadgetSuite) TestUpdateGadgetCommandlineWithNewArgs(c *C) {
+	// no command line arguments prior to the gadget update
+	bootloader.Force(s.managedbl)
+	s.state.Lock()
+	s.setupUC20ModelWithGadget(c, "pc", opts.grade)
+	s.mockModeenvForMode(c, "run")
+	devicestate.SetBootOkRan(s.mgr, true)
+	s.state.Set("seeded", true)
+
+	// mimic system state
+	m, err := boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	m.CurrentKernelCommandLines = []string{
+		"snapd_recovery_mode=run console=ttyS0 console=tty1 panic=-1",
+	}
+	c.Assert(m.Write(), IsNil)
+	err = s.managedbl.SetBootVars(map[string]string{
+		"snapd_extra_cmdline_args": "",
+	})
+	c.Assert(err, IsNil)
+	s.managedbl.SetBootVarsCalls = 0
+
+	s.state.Unlock()
+
+	yaml := gadgetYaml + `
+kernel-cmdline:
+  allow:
+    - par1=val
+    - par2
+    - append1=val
+    - append2
+`
+	files := [][]string{
+		{"meta/gadget.yaml", yaml},
+	}
+	expLog := ""
+	if opts.updated {
+		expLog = "Updated kernel command line"
+	}
+	if opts.notAllowedCmdline != "" {
+		expLog = fmt.Sprintf("%q is not allowed by the gadget and has been filtered out from the kernel command line", opts.notAllowedCmdline)
+	}
+	// The task comes from setting a system option so it is not a
+	// real gadget update and to/from files are the same.
+	s.testGadgetCommandlineUpdateRun(c, files, files, "", expLog, opts)
+
+	m, err = boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+
+	// gadget arguments are picked up for the candidate command line
+	oldCmdline := "snapd_recovery_mode=run console=ttyS0 console=tty1 panic=-1"
+	newCmdline := strutil.JoinNonEmpty([]string{
+		"snapd_recovery_mode=run console=ttyS0 console=tty1 panic=-1",
+		opts.allowedCmdline}, " ")
+	if opts.grade == "dangerous" {
+		newCmdline = strutil.JoinNonEmpty([]string{
+			newCmdline, opts.cmdlineAppendDanger}, " ")
+	}
+	expCmdlines := []string{oldCmdline}
+	numSetBootVarsCalls := 0
+	// It might have not changed if all arguments are forbidden by the gadget
+	if newCmdline != oldCmdline {
+		expCmdlines = append(expCmdlines, newCmdline)
+		numSetBootVarsCalls = 1
+	}
+
+	c.Check([]string(m.CurrentKernelCommandLines), DeepEquals, expCmdlines)
+	c.Check(s.managedbl.SetBootVarsCalls, Equals, numSetBootVarsCalls)
+	vars, err := s.managedbl.GetBootVars("snapd_extra_cmdline_args")
+	c.Assert(err, IsNil)
+	// bootenv was cleared
+	extraArgs := opts.allowedCmdline
+	if opts.grade == "dangerous" {
+		extraArgs = strutil.JoinNonEmpty([]string{extraArgs, opts.cmdlineAppendDanger}, " ")
+	}
+	c.Assert(vars, DeepEquals, map[string]string{
+		"snapd_extra_cmdline_args": extraArgs,
+	})
+}
+
+func (s *deviceMgrGadgetSuite) TestUpdateGadgetCommandlineWithNewAppendedArgs(c *C) {
+	var opts testGadgetCommandlineUpdateOpts
+	for _, isClassic := range []bool{false, true} {
+		opts = testGadgetCommandlineUpdateOpts{
+			updated:        true,
+			isClassic:      isClassic,
+			grade:          "dangerous",
+			cmdlineAppend:  "append1=val append2",
+			allowedCmdline: "append1=val append2",
```

Comment:
I wonder if there is a way have a  "c.Check" or "expectedCmdline" (or somehting simialr) here, feels a bit unusal to have this all implciit inside testUpdateGadgetCommandlineWithNewAppendedArgs().